### PR TITLE
Automatically configure heat_metadata_server_url

### PR DESCRIPTION
**Pull request by bshephar**
This change adds capability to the Heat controller to detect the configured HeatCfnAPI Route and set it as the `heat_metadata_server_url` option in the `heat.conf` file.

### CODE REVIEW

**Comment by bshephar** [comment]
```
$ oc get secret heat-config-data -o yaml | ./yq '.data["heat.conf"]' | base64 -d | grep heat_metadata_server_url
heat_metadata_server_url=https://heat-cfnapi-public-openstack.apps.ostest.test.metalkube.org
```

**Comment by bshephar** [comment]
> In general from the conf description of the parameter I am wondering if we have to set it:
> 
> ```
> # URL of the Heat metadata server. NOTE: Setting this is only needed if you
> # require instances to use a different endpoint than in the keystone catalog
> # (string value)
> #heat_metadata_server_url = <None>
> ```
> 
>  ```
> sh-5.1$ openstack endpoint list |grep heat-cfn
> | 5b3ef34b3cd04a08bef96a5dfcb2d177 | regionOne | heat-cfn         | cloudformation          | True    | public    | https://heat-cfnapi-public-openstack.apps-crc.testing/v1              |
> | 5ee5fa4bf5d04e6e926fd67adfd663ef | regionOne | heat-cfn         | cloudformation          | True    | internal  | https://heat-cfnapi-internal.openstack.svc:8000/v1                    |
> ```

There's Tempest tests failing in unialpha because the VM's are receiving the internal endpoint instead of public. Obviously, they can't reach a svc address. That specific issues was resolved by setting something else, but ultimately, the whole thing got me thinking about what other possible internal endpoints might be returned. So I started checking what we were doing in TripleO and we are indeed setting this to the public endpoint:
https://github.com/openstack-archive/tripleo-heat-templates/blob/stable/wallaby/deployment/heat/heat-engine-container-puppet.yaml#L174-L178

So this PR is basically just bringing that standard from what we had as a default in TripleO.

File: controllers/heat_controller.go [hunk 1]
```diff
@@ -1289,3 +1308,27 @@ func (r *HeatReconciler) checkHeatCfnGeneration(
...
+		}
+	}
+
+	if _, ok := cf.Spec.Override.Service["public"]; !ok {
+		r.Log.Info("no public override found in HeatCfnAPI spec")
+		return "", nil
+	}
+
+	if cf.Spec.Override.Service["public"].EndpointURL != nil {
+		return *cf.Spec.Override.Service["public"].EndpointURL, nil
```
[the remark this answers is not in the document]
Comment (bshephar) [reply controllers/heat_controller.go:1330]: Na, not for this one. But for the `heat_waitcondition_server_url` one, I did:
https://github.com/openstack-k8s-operators/heat-operator/pull/381/files#diff-55d6154f9cc993e8f39d571ed7402c3b05d9764d5a806cd3edeaa8dc1164fb55R11

**Comment by stuggi** [comment]
> > In general from the conf description of the parameter I am wondering if we have to set it:
> > ```
> > # URL of the Heat metadata server. NOTE: Setting this is only needed if you
> > # require instances to use a different endpoint than in the keystone catalog
> > # (string value)
> > #heat_metadata_server_url = <None>
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```
> > sh-5.1$ openstack endpoint list |grep heat-cfn
> > | 5b3ef34b3cd04a08bef96a5dfcb2d177 | regionOne | heat-cfn         | cloudformation          | True    | public    | https://heat-cfnapi-public-openstack.apps-crc.testing/v1              |
> > | 5ee5fa4bf5d04e6e926fd67adfd663ef | regionOne | heat-cfn         | cloudformation          | True    | internal  | https://heat-cfnapi-internal.openstack.svc:8000/v1                    |
> > ```
> 
> There's Tempest tests failing in unialpha because the VM's are receiving the internal endpoint instead of public. Obviously, they can't reach a svc address. That specific issues was resolved by setting something else, but ultimately, the whole thing got me thinking about what other possible internal endpoints might be returned. So I started checking what we were doing in TripleO and we are indeed setting this to the public endpoint: https://github.com/openstack-archive/tripleo-heat-templates/blob/stable/wallaby/deployment/heat/heat-engine-container-puppet.yaml#L174-L178
> 
> So this PR is basically just bringing that standard from what we had as a default in TripleO.

ok, maybe we could configure tempest too use the public endpoints from heat when do the catalog lookup, but set the defaults as we had in TripleO is probably a good idea. Thanks for clarification

File: controllers/heat_controller.go [hunk 2]
```diff
@@ -1289,3 +1308,27 @@ func (r *HeatReconciler) checkHeatCfnGeneration(
...
+	cf := &heatv1beta1.HeatCfnAPI{}
+	cfKey := client.ObjectKey{Namespace: instance.Namespace, Name: fmt.Sprintf("%s-cfnapi", instance.Name)}
+
+	err = r.Client.Get(context.Background(), cfKey, cf)
+	if err != nil {
+		r.Log.Error(err, "unable to get HeatCfnAPI instance: %w")
+		if k8s_errors.IsNotFound(err) {
+			return "", nil
+		}
+	}
```
[the remark this answers is not in the document]
Comment (bshephar) [reply controllers/heat_controller.go]: Yeah, you're right. No need to look it up from the CFN CR. That simplifies the implementation a lot.

**Comment by bshephar** [comment]
> > > In general from the conf description of the parameter I am wondering if we have to set it:
> > > ```
> > > # URL of the Heat metadata server. NOTE: Setting this is only needed if you
> > > # require instances to use a different endpoint than in the keystone catalog
> > > # (string value)
> > > #heat_metadata_server_url = <None>
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > ```
> > > sh-5.1$ openstack endpoint list |grep heat-cfn
> > > | 5b3ef34b3cd04a08bef96a5dfcb2d177 | regionOne | heat-cfn         | cloudformation          | True    | public    | https://heat-cfnapi-public-openstack.apps-crc.testing/v1              |
> > > | 5ee5fa4bf5d04e6e926fd67adfd663ef | regionOne | heat-cfn         | cloudformation          | True    | internal  | https://heat-cfnapi-internal.openstack.svc:8000/v1                    |
> > > ```
> > 
> > 
> > There's Tempest tests failing in unialpha because the VM's are receiving the internal endpoint instead of public. Obviously, they can't reach a svc address. That specific issues was resolved by setting something else, but ultimately, the whole thing got me thinking about what other possible internal endpoints might be returned. So I started checking what we were doing in TripleO and we are indeed setting this to the public endpoint: https://github.com/openstack-archive/tripleo-heat-templates/blob/stable/wallaby/deployment/heat/heat-engine-container-puppet.yaml#L174-L178
> > So this PR is basically just bringing that standard from what we had as a default in TripleO.
> 
> ok, maybe we could configure tempest too use the public endpoints from heat when do the catalog lookup, but set the defaults as we had in TripleO is probably a good idea. Thanks for clarification

Yeah, for Tempest we just configured the heat_client to use the public endpoint. But yeah, just trying to reduce friction from the end users perspective.

**Pull request closed without merging**